Code/Source/solver/mat_models.cpp

```diff
@@ -542,523 +701,129 @@ void get_pk2cc(const ComMod& com_mod, const CepMod& cep_mod, const dmnType& lDmn
-      double c4s  = 1.0 - one_over_exp_plus_one_s;
-      
-      // Approx. derivative of smoothed heaviside function
-      //double dc4f = 0.25*stM.khs*exp(-stM.khs*abs(Eff));
-      //double dc4s = 0.25*stM.khs*exp(-stM.khs*abs(Ess));
-
-      // Exact first derivative of smoothed heaviside function (from Wolfram Alpha)
-      double dc4f = k * (one_over_exp_plus_one_f - pow(one_over_exp_plus_one_f,2));
-      double dc4s = k * (one_over_exp_plus_one_s - pow(one_over_exp_plus_one_s,2));
-
-      // Exact second derivative of smoothed heaviside function (from Wolfram Alpha)
-      double ddc4f = pow(k,2) * (-one_over_exp_plus_one_f + 3.0*pow(one_over_exp_plus_one_f,2) - 2.0*pow(one_over_exp_plus_one_f,3));
-      double ddc4s = pow(k,2) * (-one_over_exp_plus_one_s + 3.0*pow(one_over_exp_plus_one_s,2) - 2.0*pow(one_over_exp_plus_one_s,3));
-
-      // Isochoric stress and stiffness
-      double g1 = stM.a * exp(stM.b*(Inv1-3.0));
-      auto Sb = g1*IDm;
-      double r1 = J2d/nd*mat_fun::mat_ddot(C, Sb, nsd);
-
-      g1 = g1*2.0*J4d*stM.b;
-      auto CCb = g1 * ten_dyad_prod(IDm, IDm, nsd);
-
-      // Add isochoric stress and stiffness contribution
-      S = J2d*Sb - r1*Ci;
-
-      auto PP = ten_ids(nsd) - (1.0/nd) * ten_dyad_prod(Ci, C, nsd);
-      CC = ten_ddot(CCb, PP, nsd);
-      CC = ten_transpose(CC, nsd);
-      CC = ten_ddot(PP, CC, nsd);
-      CC = CC + 2.0*r1 * (ten_symm_prod(Ci, Ci, nsd) - 
-                1.0/nd * ten_dyad_prod(Ci, Ci, nsd)) 
-              - 2.0/nd * (ten_dyad_prod(Ci, S, nsd) + ten_dyad_prod(S, Ci, nsd));
-
-      // Now add aniostropic components
-      // Fiber-sheet interaction terms
-      g1 = 2.0 * stM.afs * exp(stM.bfs*Efs*Efs);
-      auto Hfs = mat_symm_prod(fl.col(0), fl.col(1), nsd);
-      S  = S + g1*Efs*Hfs;
-
-      g1 = g1 * 2.0*(1.0 + 2.0*stM.bfs*Efs*Efs);
+    // Number of spatial dimensions
+    int nsd = com_mod.nsd;
+
+    if (nsd == 2) {
+        // Copy deformation gradient to Eigen matrix
+        auto F_2D = mat_fun::toEigenMatrix<Eigen::Matrix2d>(F);
+        
+        // Copy fiber directions to Eigen matrix
+        Eigen::Matrix<double, 2, Eigen::Dynamic> fl_2D(2, nfd);
+        for (int i = 0; i < nfd; i++) {
+            fl_2D(0, i) = fl(0, i);
+            fl_2D(1, i) = fl(1, i);
+        }
 
-      CC = CC + g1*ten_dyad_prod(Hfs, Hfs, nsd);
+        // Initialize stress and elasticity tensors
+        Eigen::Matrix2d S_2D = Eigen::Matrix2d::Zero();
+        Eigen::Matrix4d Dm_2D = Eigen::Matrix4d::Zero();
 
-      // Fiber-fiber interaction stress + additional reinforcement (Tfa)
-      double rexp = exp(stM.bff * Eff * Eff);
-      g1 = c4f*Eff*rexp;
-      g1 = g1 + (0.5*dc4f/stM.bff)*(rexp - 1.0);
-      g1 = (2.0*stM.aff*g1) + Tfa;
-      auto Hff = mat_dyad_prod(fl.col(0), fl.col(0), nsd);
-      S  = S + g1*Hff;
+        // Call templated function
+        _get_pk2cc<2>(com_mod, cep_mod, lDmn, F_2D, nfd, fl_2D, ya, S_2D, Dm_2D, Ja);
```

Comment:
Note that a function with a prefixed  `_` has a special meaning in the C++ implementation (e.g. global namespaces).

Reply:
See my response to your other comment about the meaning of underscore here.